designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentService.scala

```diff
@@ -186,72 +97,65 @@ class DefaultComponentService private(config: Config,
+      .map { subprocesses =>
+        val componentObjects = componentObjectsService.prepare(processingType, processingTypeData, user, subprocesses)
+        createComponents(componentObjects, componentUsages, processingType, componentIdProvider)
+      }
+  }
 
-            ComponentListElement(
-              id = componentId,
-              name = com.label,
-              icon = icon,
-              componentType = com.`type`,
-              componentGroupName = group.name,
-              categories = com.categories,
-              links = links,
-              usageCount = usageCount
-            )
-          }))
-      })
+  private def createComponents(componentObjects: ComponentObjects,
+                               componentUsages: Map[ComponentId, Long],
+                               processingType: ProcessingType,
+                               componentIdProvider: ComponentIdProvider): List[ComponentListElement] = {
+    componentObjects
+      .templates
+      .map { case (groupName, com) =>
+        val componentId = componentIdProvider.createComponentId(processingType, com.label, com.`type`)
+        val icon = getComponentIcon(componentObjects.config, com)
+        val links = createComponentLinks(componentId, com, componentObjects.config)
+        val usageCount = componentUsages.getOrElse(componentId, 0L)
+
+        ComponentListElement(
+          id = componentId,
+          name = com.label,
+          icon = icon,
+          componentType = com.`type`,
+          componentGroupName = groupName,
+          categories = com.categories,
+          links = links,
+          usageCount = usageCount
+        )
+      }
   }
 
-  private def getComponentUsages(categories: List[Category])(implicit loggedUser: LoggedUser, ec: ExecutionContext): Future[Map[ComponentId, Long]] = {
-    processService
-      .getProcesses[DisplayableProcess](loggedUser)
-      .map(_.filter(p => categories.contains(p.processCategory))) //TODO: move it to service?
-      .map(processes => ProcessObjectsFinder.computeComponentsUsageCount(componentIdProvider, processes))
+  private def createComponentLinks(componentId: ComponentId,
+                                   component: ComponentTemplate,
+                                   componentsConfig: ComponentsUiConfig): List[ComponentLink] = {
+    val componentLinks = componentLinksConfig
+      .filter(_.isAvailable(component.`type`))
+      .map(_.toComponentLink(componentId, component.label))
+
+    //If component configuration contains documentation link then we add base link
+    getComponentDoc(componentsConfig, component)
+      .map(ComponentLink.createDocumentationLink)
+      .map(doc => List(doc) ++ componentLinks)
+      .getOrElse(componentLinks)
   }
 
-  private def deduplication(components: Iterable[ComponentListElement]) = {
+  private def deduplicateById(components: Iterable[ComponentListElement]): List[ComponentListElement] = {
```

Comment:
Let's rename it to collect usages across processing types or something. Also it would be more visible what happens here if it would return component id -> usageCount. Let's extract this "taking first element logic" to the one place where we will describe why it is safe to skip other elements.

Reply:
Renamed to `mergeSameComponentsAcrossProcessingTypes`.